Fix for 4th gen Chromecast

iMacThere4iAm opened the pull request:
On newer Chromecast devices which don't return the `videoID` to `go-chromecast`, attempt to identify it using an API search query. Not guaranteed to get it right.
Requires an API key.
Fixes [nichobi/sponsorblockcast#16](https://github.com/nichobi/sponsorblockcast/issues/16)

TODO: Needs error handling, in case an API query is attempted but fails.
TODO: Needs some validation to confirm back that it's identified the right `videoID`. Perhaps by comparing the duration returned by `go-chromecast` to the `videoDuration` in the `sponsorblock` database.
TODO: Could scrape a non-API Youtube search as a secondary fallback option if the API query fails. This would be less robust but doesn't require an API key.

iMacThere4iAm committed:
On newer Chromecast devices which don't return the videoID to go-chromecast, attempt to identify it using an API search query. Not guaranteed to get it right.
Requires an API key.
TODO: Needs error handling, in case an API query is attempted but fails.
TODO: Needs some validation to confirm back that it's identified the right videoID. Perhaps by comparing the duration returned by go-chromecast to the videoDuration in the sponsorblock database.
TODO: Could scrape a non-API Youtube search as a secondary fallback option if the API query fails. This would be less robust but doesn't require an API key.

iMacThere4iAm committed:
Fix a logic eror that prevents the script from detecting that the currently playing video has changed.

iMacThere4iAm committed:
Use --data syntax for curl which is slightly more readable.

nichobi commented on README.md:
```diff
@@ -31,14 +31,15 @@ First you will need a `docker-compose.yaml` file, such as the example included.
 
 ## Usage
-Run `sponsorblockcast` from a terminal or activate the service with `systemd enable --now sponsorblockcast`
+Run `sponsorblockcast` from a terminal or activate the service with `systemd enable sponsorblockcast && systemctl start sponsorblockcast`
```
Any particular reason for this change besides being more explicit?

nichobi replied:
Also, noticing now I'd written the command as `systemd` instead of `systemctl`, oops!

iMacThere4iAm replied:
The `--now` option wouldn't work for me on Raspberry Pi OS. 

I'm newly learning to use `git` and realised too late that these changes should have been in separate commits.

iMacThere4iAm replied:
Unless the `systemd` typo was the real reason why it didn't work! I'll re-test.

iMacThere4iAm replied:
That was it all along. 50e2caf

nichobi commented on sponsorblockcast.sh:
```diff
@@ -28,6 +32,19 @@ watch () {
       video_id=$(echo "$status" | grep -oP "id=\"\K[^\"]+")
+      video_title=$(echo "$status" | grep -oP "title=\"\K[^\"]+")
+      video_artist=$(echo "$status" | grep -oP "artist=\"\K[^\"]+")
+      if [ -z "$video_id" ]
```
Check if the API key is defined here too, else there's no point calling the API

iMacThere4iAm replied:
Good call. 531a48ac2c95aaa229d3272935cc5078f2090b5d

nichobi reviewed (commented):
Thanks for this! Looks promising so far, got a few comments here and there, nothing major. Appreciate the cleanup of the curl command and the addition of traps!

Some of the changes might be better off in separate commits, like the signal traps and the changed `systemctl` commands in the readme, as they aren't connected to this feature.

iMacThere4iAm committed:
Revert to the --now syntax and correct the original typo.

iMacThere4iAm committed:
Recommend setting the API key environment variable in the systemd override.

iMacThere4iAm committed:
Only attempt fallback search if the API key is set.

iMacThere4iAm commented:
This fix needs more testing, at the moment it works well when run from a terminal but not via systemd.

iMacThere4iAm committed:
Don't ignore $SBCPOLLINTERVAL.

iMacThere4iAm committed:
Improve API search accuracy using intitle: operator and escaped quotes.

iMacThere4iAm committed:
Use a separate function to test for changed $video_title in the watch() loop.

This should make the logic easier to follow and debug by eliminating if statements.

iMacThere4iAm committed:
Move API search to a separate function.

iMacThere4iAm committed:
Try to get segments only if the videoID is known.

iMacThere4iAm committed:
Parse the segments file on load instead of on save

This keeps useful data such as duration, which can be used to validate the API search match.
Save the segments file as "$id".json instead of "$id".segments, since it is now valid JSON.

iMacThere4iAm committed:
Add comments

iMacThere4iAm committed:
Fix malformed search URL

iMacThere4iAm committed:
Fix for curl/curl#6200

Workaround for curl version 7.74.0 and earlier which would fail with an "Out of Memory" error if $HOME is unset.
It happens that the systemd environment does not set $HOME by default so adding it explicitly here avoids the bug.

iMacThere4iAm committed:
Don't try to match when an ad is playing

If an ad is playing, `go-chromecast watch` returns "YouTube (PLAYING), unknown" which is no help for matching the VideoID.
This fix avoids trying to match until a proper video is playing.

iMacThere4iAm committed:
Add $SBCDEBUG environment variable for easier debugging under systemd

iMacThere4iAm committed:
Give feedback on whether skippable segments exist for the current video

iMacThere4iAm commented:
> This fix needs more testing, at the moment it works well when run from a terminal but not via systemd.

Fixed by a04b1a5. Everything else is just tinkering.

iMacThere4iAm commented:
Pull request closed by accident!

nichobi committed:
Minor code formatting

nichobi committed:
Move video_title/artist to within if statement

Avoids grepping for variables when they won't be used.

nichobi committed:
Add shellcheckrc to disable single quote warning

nichobi commented:
Looks great! I've made a few formatting changes (just being picky, nothing wrong with it) and a few minor changes. If it all looks good to you, I'm happy to merge it now!

iMacThere4iAm commented:
All good, be as picky as you like with my noob code! I've got nothing to add to the branch, so merge away.
